File: backends/kotlin-ktor/src/main/kotlin/com/authorization/domain/invitation/AuthRepository.kt
/*
 * 招待認証リポジトリインターフェースモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.domain.invitation

/**
 * 招待認証トークンのキャッシュリポジトリインターフェースです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
interface AuthRepository {
    /** トークンとロールを指定秒数キャッシュします。 */
    suspend fun store(token: String, role: Int, ttl: Long)
    /** トークンに紐づくロールを取得します。存在しない場合は null を返します。 */
    suspend fun find(token: String): Int?
    /** トークンを削除します。 */
    suspend fun remove(token: String)
}

File: backends/rust-axum/src/domain/invitation/auth_repository.rs
//! 招待認証リポジトリインターフェースモジュール。
//!
//! # Author
//! Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

use async_trait::async_trait;

pub type DomainError = Box<dyn std::error::Error + Send + Sync>;

/// 招待認証トークンのキャッシュリポジトリインターフェース。
#[async_trait]
pub trait AuthRepository: Send + Sync {
    /// トークンとロールを指定秒数キャッシュします。
    async fn store(&self, token: &str, role: u8, ttl: u64) -> Result<(), DomainError>;
    /// トークンに紐づくロールを取得します。存在しない場合は None を返します。
    async fn find(&self, token: &str) -> Result<Option<u8>, DomainError>;
    /// トークンを削除します。
    async fn remove(&self, token: &str) -> Result<(), DomainError>;
}
